src/dune_engine/sandbox.ml

```diff
@@ -295,13 +295,25 @@ let move_targets_to_build_dir t ~loc ~should_be_skipped ~(targets : Targets.Vali
       Path.Build.Set.to_list_map targets.dirs ~f:(fun target ->
         let src_dir = map_path t target in
         let files = collect_dir_recursively ~loc ~src_dir ~dst_dir:target in
-        if Path.Untracked.exists (Path.build target)
-        then
-          (* We clean up all targets (including directory targets) before running an
-             action, so this branch should be unreachable. *)
-          Code_error.raise
-            "Stale directory target in the build directory"
-            [ "dst_dir", Path.Build.to_dyn target ];
+        (match Path.Untracked.stat (Path.build target) with
+         | Error (Unix.ENOENT, _, _) -> ()
+         | Error e ->
+           User_error.raise
+             ~hints:
+               [ Pp.text
+                   "delete this file manually or check the permissions of the parent \
+                    directory of this file"
+               ]
+             [ Pp.textf "unable to stat %s" (Path.Build.to_string_maybe_quoted target)
+             ; Pp.text "reason:"
+             ; Pp.text (Unix_error.Detailed.to_string_hum e)
+             ]
+         | Ok { Unix.st_kind; _ } ->
+           (* We clean up all targets (including directory targets) before running an
+              action, so this branch should be unreachable. *)
+           Code_error.raise
+             "Stale directory target in the build directory"
```

Comment:
Should this be a Code error? I feel like this is an issue poorly behaved rules can introduce.

Reply:
Can you demonstrate that with a test?

Reply:
We are hitting this error when building the compiler, for some reason the sandbox is still around. I still don't fully understand the issue so we haven't been able to make a minimal reproduction yet.

Reply:
A test has been added. I also changed it to a User_error because otherwise the test isn't reproducible.

Reply:
Thanks! The test build doesn't seem to fail though?

Reply:
I think its failing correctly in the CI, just isn't updated.

Reply:
Ah yep. It's updated now.